### PR TITLE
[AMQ-9530]Fix SelectorAwareVirtualTopicInterceptor ClassCastException if next is not Topic.

### PR DESCRIPTION
**Pull request by Nikita-Shupletsov**
`next` is not guaranteed  to be Topic. If there is plugin that injects another `DestinationInterceptor`, the code will fail with `ClassCastException`.

Added a method to correctly unwrap next.

Ran:
* AMQ4899Test.java
* VirtualTopicSelectorTest.java
* JavaVirtualDestTest.java
* SelectorAwareVTThatDropsMessagesWhenNoConsumer.java
* TwoBrokerVirtualTopicSelectorAwareForwardingTest.java
* VirtualConsumerDemandTest.java
* VirtualTopicSelectorWithAnotherInterceptorTest.java
to make sure there is no regression

### CODE REVIEW

**Comment by mattrpav** [comment]
What is the use case for this?

**Comment by Nikita-Shupletsov** [comment]
> What is the use case for this?

you mean the use case of having a DestinationFilter after SelectorAwareVirtualTopicInterceptor?

I my case I add an interceptor to all topics. and that interceptor coexists with the rest of interceptors. and some topics can be vistual topics. it's technically possible on the plugin side to check if there is a virtual topic interceptor and not add my interceptor. But I checked other interceptors([this one](https://github.com/apache/activemq/blob/main/activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/MappedQueueFilter.java#L68) for example) and they handle this situation correctly, so I thought it would be a positive change in general. In case people decide to do something similar in the future, they will not need to work around the edge case.

**Review by kenliao94 — APPROVED** [review]

**Comment by mattrpav** [comment]
A couple of points:

1. This change needs to be a JIRA
2. This needs a unit test with an example configured w/ another DestinationFilter on top of Virtual Topic
3. If we are going to fix it, we should adapt all the Virtual Destinations to use the same set of methods.

**Comment by Nikita-Shupletsov** [comment]
>     1. This change needs to be a JIRA
Done. https://issues.apache.org/jira/browse/AMQ-9530
>     2. This needs a unit test with an example configured w/ another DestinationFilter on top of Virtual Topic
Updated `VirtualTopicSelectorTest,` as it's basically tests `SelectorAwareVirtualTopicInterceptor`. Also used it to initially reproduce the issue(see the ticket).
>     3. If we are going to fix it, we should adapt all the Virtual Destinations to use the same set of methods.
There are no issues like this in other pars of the Virtual Destination code. only in `SelectorAwareVirtualTopicInterceptor`

**Comment by mattrpav** [comment]
> There are no issues like this in other pars of the Virtual Destination code. only in `SelectorAwareVirtualTopicInterceptor`

MappedQueueFilter has a nearly identical method:

ref: https://github.com/apache/activemq/blob/5309352e12c1d5f05c027331f49cc1caa6a62f2f/activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/MappedQueueFilter.java#L90

**Comment by Nikita-Shupletsov** [comment]
> > There are no issues like this in other pars of the Virtual Destination code. only in `SelectorAwareVirtualTopicInterceptor`
> 
> MappedQueueFilter has a nearly identical method:
> 
> ref:
> 
> https://github.com/apache/activemq/blob/5309352e12c1d5f05c027331f49cc1caa6a62f2f/activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/MappedQueueFilter.java#L90

are you proposing to extract this method to a helper class? If I create a new class, let's say VirtualTopicSupport or VistualTopicHelper, will it be okay? or do you have a particular place in mind fir it?

**Comment by mattrpav** [comment]
Yes, a base class that contains the lookup method to unwrap the destination to get to the root (real) destination that can be used by VirtualTopic and MappedQueueFilter.

ie BaseVirtualDestination(Inteceptor|Filter)

**Comment by Nikita-Shupletsov** [comment]
> Yes, a base class that contains the lookup method to unwrap the destination to get to the root (real) destination that can be used by VirtualTopic and MappedQueueFilter.
> 
> ie BaseVirtualDestination(Inteceptor|Filter)

sounds good! done

File: activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.Optional;
+
+public class BaseVirtualDestinationFilter extends DestinationFilter {
+    
+    public BaseVirtualDestinationFilter(Destination next) {
+        super(next);
+    }
+
+    Optional<BaseDestination> getBaseDestination(Destination virtualDest) {
```
[the remark this answers is not in the document]
Comment (cshannon) [reply activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java:15]: I would probably avoid optional if it's a hot spot area of the code and just do a null check.

File: activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package org.apache.activemq.broker.region.virtual;
```
Comment (jbonofre) [inline activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java:1]: The ASF header is missing here.

File: activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbonofre) [reply activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java:15]: I don't think we should use `Optional` here. I would rather check the base object.

**Review by jbonofre — CHANGES_REQUESTED** [review]

File: activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nikita-Shupletsov) [reply activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java]: Done!

**Review by jbonofre — APPROVED** [review]
The change looks good to me now.

@mattrpav as you requested changes, can you please take a new look ?

File: activemq-unit-tests/src/test/java/org/apache/activemq/broker/virtual/VirtualTopicSelectorTest.java [hunk 3]
```diff
@@ -99,7 +103,25 @@ protected BrokerService createBroker() throws Exception {
         virtualTopic.setSelectorAware(true);
         VirtualDestinationInterceptor interceptor = new VirtualDestinationInterceptor();
         interceptor.setVirtualDestinations(new VirtualDestination[]{virtualTopic});
-        broker.setDestinationInterceptors(new DestinationInterceptor[]{interceptor});
+        TestDestinationInterceptor testInterceptor = new TestDestinationInterceptor();
```
Comment (mattrpav) [inline activemq-unit-tests/src/test/java/org/apache/activemq/broker/virtual/VirtualTopicSelectorTest.java:106]: Instead of modifying the existing test, please add a second test to cover the new use case. Changing the test removes the test for the out-of-the-box default behavior vs the custom behavior this fix solves.

**Review by mattrpav — CHANGES_REQUESTED** [review]
The code changes look good.

o

File: activemq-unit-tests/src/test/java/org/apache/activemq/broker/virtual/VirtualTopicSelectorTest.java (hunk 3 above)
Comment (jbonofre) [reply activemq-unit-tests/src/test/java/org/apache/activemq/broker/virtual/VirtualTopicSelectorTest.java:106]: @Nikita-Shupletsov do you mind to add a new test as proposed by @mattrpav ? I would like to include this PR in ActiveMQ 6.1.4 release.
Thanks !

Comment (jbonofre) [reply activemq-unit-tests/src/test/java/org/apache/activemq/broker/virtual/VirtualTopicSelectorTest.java]: It looks good to me. Thanks !

Comment (jbonofre) [reply activemq-unit-tests/src/test/java/org/apache/activemq/broker/virtual/VirtualTopicSelectorTest.java]: We could have create an abstract but it's really cosmetic, let's go like this for now.

**Comment by jbonofre** [comment]
FYI, in some contexts, the `VirtualConsumerDemandTest` can be stuck. It's not related to this PR (known issue).

**Comment by mattrpav** [comment]
This type of internal broker plugin and extension points (API/SPI) change should happen in a minor release 6.2.x and not a minor release.

**Comment by jbonofre** [comment]
@mattrpav as it's a bug fix, I think it's ok for have it on all active branches.

File: activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.broker.region.virtual;
+
+import org.apache.activemq.broker.region.BaseDestination;
+import org.apache.activemq.broker.region.Destination;
+import org.apache.activemq.broker.region.DestinationFilter;
+
+import java.util.Optional;
```
[the remark this answers is not in the document]
Comment (Nikita-Shupletsov) [reply activemq-broker/src/main/java/org/apache/activemq/broker/region/virtual/BaseVirtualDestinationFilter.java:23]: Addressed here: https://github.com/apache/activemq/pull/1351
